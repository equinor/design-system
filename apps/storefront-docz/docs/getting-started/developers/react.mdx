---
title: Developers
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
