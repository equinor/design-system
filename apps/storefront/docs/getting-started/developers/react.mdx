---
title: Developers
metaTitle: Developers react
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
