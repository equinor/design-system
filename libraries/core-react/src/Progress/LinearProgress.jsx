import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import styled, { css, keyframes } from 'styled-components'
import { progress as tokens } from './Progress.tokens'

const indeterminate1 = keyframes`
  0% {
        left: -35%;
        right: 100%;
      }

      60% {
        left: 100%;
        right: -90%;
      }
      100% {
        left: 100%;
        right: -90%;
      }
`
const indeterminate2 = keyframes`
   0%{
        left: -200%;
        right: 100%;
      }
      60% {
        left: 107%;
        right: -8%;
      }

      100% {
        left: 107%;
        right: -8%;
      }
    
`

const ProgressRoot = styled.div`
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: ${tokens.linear.background};
  width: 100%;
`

const ProgressBar = styled.div`
  ${({ variant }) =>
    variant === 'indeterminate'
      ? css`
          width: auto;
          animation: ${indeterminate1} 2.1s
            cubic-bezier(0.65, 0.815, 0.735, 0.395) infinite;
        `
      : css`
          transition: transform 0.4s linear;
          background-color: ${tokens.linear.overlay};
        `}
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  ${(transform) => transform};
`
const IndeterminateProgress = styled.div`
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: ${tokens.linear.overlay};
  animation: ${indeterminate2} 2.1s cubic-bezier(0.165, 0.84, 0.44, 1) 1.15s
    infinite;
`

const LinearProgress = forwardRef(function LinearProgress(
  { children, variant, className, value, ...props },
  ref,
) {
  const rootProps = {}
  let barStyle
  if (variant === 'determinate') {
    if (value !== undefined) {
      rootProps['aria-valuenow'] = Math.round(value)
      rootProps['aria-valuemin'] = 0
      rootProps['aria-valuemax'] = 100
      const transform = value - 100

      barStyle = `translateX(${transform}%)`
    }
  }

  const progressProps = {
    variant,
    transform: barStyle,
  }

  return (
    <ProgressRoot {...rootProps} role="progressbar">
      <ProgressBar className="progressBar" {...progressProps} />
      {variant === 'indeterminate' && (
        <IndeterminateProgress className="indeterminate" />
      )}
    </ProgressRoot>
  )
})

LinearProgress.displayName = 'eds-linear-progress'

LinearProgress.propTypes = {
  /** @ignore */
  className: PropTypes.string,
  /* Variant
   * Use indeterminate when there is no progress value */
  variant: PropTypes.oneOf(['determinate', 'indeterminate']),
  /* The value of the progress indicator for determinate variant
   * Value between 0 and 100 */
  value: PropTypes.number,
}

LinearProgress.defaultProps = {
  className: '',
  variant: 'indeterminate',
  value: null,
}

export { LinearProgress }
