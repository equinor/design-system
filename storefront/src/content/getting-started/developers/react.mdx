---
title: React
tabs:
    - Introduction
    - CSS framework
    - React
    - Figma
    - Contributing
mode: draft
---


:warning: Information coming soon
