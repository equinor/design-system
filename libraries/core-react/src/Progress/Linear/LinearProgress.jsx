import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import styled, { css, keyframes } from 'styled-components'
import { progress as tokens } from '../Progress.tokens'

const indeterminate = keyframes`
  0%{
    left: -200%;
    right: 100%;
  }
  60% {
    left: 107%;
    right: -8%;
  }

  100% {
    left: 107%;
    right: -8%;
  }
    
`

const ProgressRoot = styled.div`
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: ${tokens.linear.background};
  width: 100%;
  border-radius: 50px;
`

const ProgressBar = styled.div`
  ${({ variant }) =>
    variant === 'determinate' &&
    css`
      transition: transform 0.4s linear;
      background-color: ${tokens.linear.overlay};
    `}
  width: 100%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
`
const IndeterminateProgress = styled.div`
  width: 75%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: ${tokens.linear.overlay};
  animation: ${indeterminate} 1.5s cubic-bezier(0.165, 0.84, 0.44, 1) 1s
    infinite;
`

const LinearProgress = forwardRef(function LinearProgress(
  { variant, className, value, ...props },
  ref,
) {
  const rootProps = {
    ...props,
    ref,
  }
  let barStyle
  if (variant === 'determinate') {
    if (value !== undefined) {
      rootProps['aria-valuenow'] = Math.round(value)
      rootProps['aria-valuemin'] = 0
      rootProps['aria-valuemax'] = 100
      const transform = value - 100

      barStyle = `translateX(${transform}%)`
    }
  }

  const progressProps = {
    variant,
  }

  return (
    <ProgressRoot
      {...rootProps}
      role="progressbar"
      className={`${className} ${variant}-progress`}
    >
      <ProgressBar {...progressProps} style={{ transform: barStyle }} />
      {variant === 'indeterminate' && <IndeterminateProgress />}
    </ProgressRoot>
  )
})

LinearProgress.displayName = 'eds-linear-progress'

LinearProgress.propTypes = {
  /** @ignore */
  className: PropTypes.string,
  /* Variant
   * Use indeterminate when there is no progress value */
  variant: PropTypes.oneOf(['determinate', 'indeterminate']),
  /* The value of the progress indicator for determinate variant
   * Value between 0 and 100 */
  value: PropTypes.number,
}

LinearProgress.defaultProps = {
  className: '',
  variant: 'indeterminate',
  value: null,
}

export { LinearProgress }
