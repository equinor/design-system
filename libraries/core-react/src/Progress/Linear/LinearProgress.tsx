import * as React from 'react'
import { forwardRef, HTMLAttributes } from 'react'
import styled, { css, keyframes } from 'styled-components'
import type { CSSObject } from 'styled-components'

import { progress as tokens } from '../Progress.tokens'

const indeterminate = keyframes`
  0%{
    left: -200%;
    right: 100%;
  }
  60% {
    left: 107%;
    right: -8%;
  }

  100% {
    left: 107%;
    right: -8%;
  }

`

const ProgressRoot = styled.div`
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: ${tokens.linear.background};
  width: 100%;
  border-radius: 50px;
`
type StyledProgressBarProps = {
  variant: 'indeterminate' | 'determinate'
}

const ProgressBar = styled.div<StyledProgressBarProps>`
  ${({ variant }) =>
    variant === 'determinate' &&
    css`
      transition: transform 0.4s linear;
      background-color: ${tokens.linear.overlay};
    `}
  width: 100%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
`
const IndeterminateProgress = styled.div`
  width: 75%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: ${tokens.linear.overlay};
  animation: ${indeterminate} 1.5s cubic-bezier(0.165, 0.84, 0.44, 1) 1s
    infinite;
`

export type LinearProgressProps = {
  /** Variant
   * Use indeterminate when there is no progress value */
  variant?: 'indeterminate' | 'determinate'
  /** The value of the progress indicator for determinate variant
   * Value between 0 and 100 */
  value?: number
  /** @ignore */
  className?: string
} & HTMLAttributes<HTMLDivElement>

const LinearProgress = forwardRef<HTMLDivElement, LinearProgressProps>(
  function LinearProgress(
    { variant = 'indeterminate', className = '', value = null, ...props },
    ref,
  ) {
    const rootProps = {
      ...props,
      ref,
    }
    let barStyle: string
    if (variant === 'determinate') {
      if (value !== undefined) {
        rootProps['aria-valuenow'] = Math.round(value)
        rootProps['aria-valuemin'] = 0
        rootProps['aria-valuemax'] = 100
        const transform = value - 100

        barStyle = `translateX(${transform}%)`
      }
    }

    const progressProps = {
      variant,
    }

    const transformStyle: CSSObject = {
      transform: barStyle,
    }

    return (
      <ProgressRoot
        {...rootProps}
        role="progressbar"
        className={`${className} ${variant}-progress`}
      >
        <ProgressBar {...progressProps} style={transformStyle} />
        {variant === 'indeterminate' && <IndeterminateProgress />}
      </ProgressRoot>
    )
  },
)

// LinearProgress.displayName = 'eds-linear-progress'

export { LinearProgress }
