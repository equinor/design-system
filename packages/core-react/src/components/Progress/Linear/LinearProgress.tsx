import { forwardRef, HTMLAttributes } from 'react'
import styled, { keyframes } from 'styled-components'
import type { CSSObject } from 'styled-components'
import * as tokens from './LinearProgress.tokens'

const indeterminate = keyframes`
  0%{
    left: -200%;
    right: 100%;
  }
  60% {
    left: 107%;
    right: -8%;
  }

  100% {
    left: 107%;
    right: -8%;
  }

`

const Track = styled.div`
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: ${tokens.primary.background};
  width: 100%;
  border-radius: 50px;
`

const ProgressBar = styled.div`
  transition: transform 0.4s linear;
  background-color: ${tokens.primary.entities.progress.background};
  width: 100%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
`
const IndeterminateProgressBar = styled.div`
  width: 75%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: ${tokens.primary.entities.progress.background};
  animation: ${indeterminate} 1.5s cubic-bezier(0.165, 0.84, 0.44, 1) 1s
    infinite;
`

export type LinearProgressProps = {
  /** Variant
   * Use indeterminate when there is no progress value */
  variant?: 'indeterminate' | 'determinate'
  /** The value of the progress indicator for determinate variant
   * Value between 0 and 100 */
  value?: number
} & HTMLAttributes<HTMLDivElement>

const LinearProgress = forwardRef<HTMLDivElement, LinearProgressProps>(
  function LinearProgress(
    { variant = 'indeterminate', value = null, ...rest },
    ref,
  ) {
    const props = {
      ...rest,
      ref,
    }
    let barStyle: string
    if (variant === 'determinate') {
      if (value !== undefined) {
        props['aria-valuenow'] = Math.round(value)
        props['aria-valuemin'] = 0
        props['aria-valuemax'] = 100
        const transform = value - 100

        barStyle = `translateX(${transform}%)`
      }
    }

    const transformStyle: CSSObject = {
      transform: barStyle,
    }

    return (
      <Track {...props} role="progressbar">
        {variant === 'indeterminate' ? (
          <IndeterminateProgressBar />
        ) : (
          <ProgressBar style={transformStyle} />
        )}
      </Track>
    )
  },
)

// LinearProgress.displayName = 'eds-linear-progress'

export { LinearProgress }
