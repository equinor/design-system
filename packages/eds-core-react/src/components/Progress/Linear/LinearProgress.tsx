import { forwardRef, HTMLAttributes, useEffect, useState } from 'react'
import styled, { keyframes } from 'styled-components'
import type { CSSObject } from 'styled-components'
import * as tokens from './LinearProgress.tokens'

const indeterminate = keyframes`
  0%{
    left: -200%;
    right: 100%;
  }
  60% {
    left: 107%;
    right: -8%;
  }

  100% {
    left: 107%;
    right: -8%;
  }

`

const Track = styled.div`
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: ${tokens.primary.background};
  width: 100%;
  border-radius: 50px;
`

const ProgressBar = styled.div`
  transition: transform 0.4s linear;
  background-color: ${tokens.primary.entities.progress.background};
  width: 100%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
`
const IndeterminateProgressBar = styled.div`
  width: 75%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: ${tokens.primary.entities.progress.background};
  animation: ${indeterminate} 1.5s cubic-bezier(0.165, 0.84, 0.44, 1) 1s
    infinite;
`

const SrOnlyOutput = styled.output`
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
`

export type LinearProgressProps = {
  /** Variant
   * Use indeterminate when there is no progress value */
  variant?: 'indeterminate' | 'determinate'
  /** The value of the progress indicator for determinate variant
   * Value between 0 and 100 */
  value?: number
} & HTMLAttributes<HTMLDivElement>

const LinearProgress = forwardRef<HTMLDivElement, LinearProgressProps>(
  function LinearProgress(
    { variant = 'indeterminate', value = null, ...rest },
    ref,
  ) {
    const props = {
      ...rest,
      ref,
    }
    const [srProgress, setSrProgress] = useState(0)

    let barStyle: string
    if (variant === 'determinate') {
      if (value !== undefined) {
        props['aria-valuenow'] = Math.round(value)
        props['aria-valuemin'] = 0
        props['aria-valuemax'] = 100
        const transform = value - 100

        barStyle = `translateX(${transform}%)`
      }
    }

    const transformStyle: CSSObject = {
      transform: barStyle,
    }

    useEffect(() => {
      if (variant === 'indeterminate') return
      if (value >= 25 && value < 50) {
        setSrProgress(25)
      } else if (value >= 50 && value < 75) {
        setSrProgress(50)
      } else if (value >= 75 && value < 100) {
        setSrProgress(75)
      } else if (value === 100) {
        setSrProgress(100)
      }
    }, [value, variant])

    const getProgressFormatted = () => {
      return `Loading ${srProgress}%`
    }

    return (
      <>
        <Track {...props} role="progressbar">
          {variant === 'indeterminate' ? (
            <IndeterminateProgressBar />
          ) : (
            <ProgressBar style={transformStyle} />
          )}
        </Track>
        {variant === 'determinate' && (
          <SrOnlyOutput>{getProgressFormatted()}</SrOnlyOutput>
        )}
      </>
    )
  },
)

// LinearProgress.displayName = 'eds-linear-progress'

export { LinearProgress }
