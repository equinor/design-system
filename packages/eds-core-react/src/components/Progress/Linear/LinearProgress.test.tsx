/* eslint-disable no-undef */
import { render, cleanup, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { axe } from 'jest-axe'
import 'jest-styled-components'
import styled from 'styled-components'
import { LinearProgress } from './LinearProgress'

const StyledProgress = styled(LinearProgress)`
  position: absolute;
`

afterEach(cleanup)

describe('LinearProgress ', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<LinearProgress />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<LinearProgress aria-label="Label text" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('has coorect aria values when variant is "determinate"', () => {
    render(<LinearProgress variant="determinate" value={50} />)
    const progress = screen.getByRole('progressbar')
    expect(progress).toHaveAttribute('aria-valuenow', '50')
    expect(progress).toHaveAttribute('aria-valuemin', '0')
    expect(progress).toHaveAttribute('aria-valuemax', '100')
  })
  it('can extend the css for the component', () => {
    render(<StyledProgress />)
    const progress = screen.getByRole('progressbar')
    expect(progress).toHaveStyleRule('position', 'absolute')
  })
})
