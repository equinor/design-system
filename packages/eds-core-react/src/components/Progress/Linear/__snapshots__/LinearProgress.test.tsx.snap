// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress  Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  overflow: hidden;
  height: 4px;
  background-color: var(--eds_infographic_primary__moss_green_13, rgba(222, 237, 238, 1));
  width: 100%;
  border-radius: 50px;
}

.c1 {
  width: 75%;
  border-radius: 50px;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: var(--eds_infographic_primary__moss_green_100, rgba(0, 112, 121, 1));
  animation: gbxbMm 1.5s cubic-bezier(0.165,0.84,0.44,1) 1s infinite;
}

<div
    class="c0"
    role="progressbar"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;
